services/cargohold/test/integration/API/V4.hs

```diff
@@ -0,0 +1,282 @@
+      r4 <-
+        get (c . path loc . zUser uid . header "Asset-Token" (toByteString' tok) . noRedirect)
+          <!! const 404 === statusCode
+      let Just date' = C8.unpack <$> lookup "Date" (responseHeaders r4)
+      let utc' = parseTimeOrError False defaultTimeLocale rfc822DateFormat date' :: UTCTime
+      liftIO $ assertBool "bad date" (utc' >= utc)
+
+testSimpleTokens :: TestSignature ()
+testSimpleTokens c = do
+  uid <- liftIO $ Id <$> nextRandom
+  uid2 <- liftIO $ Id <$> nextRandom
+  -- Initial upload
+  let sets = defAssetSettings & set setAssetRetention (Just AssetVolatile)
+  let bdy = (applicationText, "Hello World")
+  r1 <-
+    uploadSimple (c . path "/assets/v3") uid sets bdy
+      <!! const 201 === statusCode
+  let loc = decodeHeader "Location" r1 :: ByteString
+  let Just ast = responseJsonMaybe @Asset r1
+  let key = view assetKey ast
+  let Just tok = view assetToken ast
+  -- No access without token from other user (opaque 404)
+  get (c . path loc . zUser uid2 . noRedirect)
+    !!! const 404 === statusCode
+  -- No access with wrong token (opaque 404)
+  get (c . path loc . zUser uid2 . header "Asset-Token" "acb123" . noRedirect)
+    !!! const 404 === statusCode
+  -- Token renewal fails if not done by owner
+  post (c . paths ["assets", "v3", toByteString' (qUnqualified key), "token"] . zUser uid2) !!! do
+    const 403 === statusCode
+    const (Just "unauthorised") === fmap label . responseJsonMaybe
+  -- Token renewal succeeds if done by owner
+  r2 <-
+    post (c . paths ["assets", "v3", toByteString' (qUnqualified key), "token"] . zUser uid)
+      <!! const 200 === statusCode
+  let Just tok' = newAssetToken <$> responseJsonMaybe r2
+  liftIO $ assertBool "token unchanged" (tok /= tok')
+  -- Download by owner with new token.
+  r3 <-
+    get (c . path loc . zUser uid . header "Asset-Token" (toByteString' tok') . noRedirect) <!! do
+      const 302 === statusCode
+      const Nothing === responseBody
+  r4 <- flip get' id =<< parseUrlThrow (C8.unpack (getHeader' "Location" r3))
+  liftIO $ do
+    assertEqual "status" status200 (responseStatus r4)
+    assertEqual "content-type should always be application/octet-stream" (Just applicationOctetStream) (getContentType r4)
+    assertEqual "token mismatch" tok' (decodeHeader "x-amz-meta-token" r4)
+    assertEqual "user mismatch" uid (decodeHeader "x-amz-meta-user" r4)
+    assertEqual "data mismatch" (Just "Hello World") (responseBody r4)
+  -- Verify access without token if the request comes from the creator.
+  get (c . path loc . zUser uid . noRedirect)
+    !!! const 302 === statusCode
+  -- Verify access with new token from a different user.
+  get (c . path loc . header "Asset-Token" (toByteString' tok') . zUser uid2 . noRedirect)
+    !!! const 302 === statusCode
+  -- Delete Token fails if not done by owner
+  delete (c . paths ["assets", "v3", toByteString' (qUnqualified key), "token"] . zUser uid2) !!! do
```

Comment:
v3 or v4?

Reply:
That's indeed a good question, because both - `v3` and `v4` - are available. :thinking:

I'll check that nothing is untested in that case.

Reply:
Deleting an asset is still `v3`.